File: common/src/main/java/generations/gg/generations/core/generationscore/world/container/GenerationsContainers.java
package generations.gg.generations.core.generationscore.world.container;

import dev.architectury.event.events.common.PlayerEvent;
import dev.architectury.registry.menu.MenuRegistry;
import dev.architectury.registry.registries.DeferredRegister;
import dev.architectury.registry.registries.RegistrySupplier;
import generations.gg.generations.core.generationscore.GenerationsCore;
import generations.gg.generations.core.generationscore.world.level.block.entities.CookingPotBlockEntity;
import generations.gg.generations.core.generationscore.world.level.block.entities.MachineBlockEntity;
import generations.gg.generations.core.generationscore.world.level.block.entities.RksMachineBlockEntity;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.flag.FeatureFlagSet;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.MenuType;
import net.minecraft.world.level.block.entity.BlockEntity;

import java.util.function.Function;

public class GenerationsContainers {
    public static final DeferredRegister<MenuType<?>> CONTAINERS = DeferredRegister.create(GenerationsCore.MOD_ID, Registries.MENU);
    public static final RegistrySupplier<MenuType<CookingPotContainer>> COOKING_POT = register("cooking_pot", CookingPotContainer::new, CookingPotBlockEntity.class);
    public static final RegistrySupplier<MenuType<GenericChestContainer>> GENERIC = CONTAINERS.register("generic", () -> MenuRegistry.ofExtended(GenericChestContainer::new));
    public static final RegistrySupplier<MenuType<MachineBlockContainer>> MACHINE_BLOCK = register("machine_block", MachineBlockContainer::new, MachineBlockEntity.class);
    public static final RegistrySupplier<MenuType<MelodyFluteContainer>> MELODY_FLUTE = CONTAINERS.register("melody_flute", () -> MenuRegistry.ofExtended(MelodyFluteContainer::new));

    public static final RegistrySupplier<MenuType<TrashCanContainer>> TRASHCAN = CONTAINERS.register("trashcan", () -> new MenuType<>(TrashCanContainer::new, FeatureFlagSet.of()));
    public static final RegistrySupplier<MenuType<WalkmonContainer>> WALKMON = CONTAINERS.register("walkmon", () -> MenuRegistry.ofExtended(WalkmonContainer::new));

    public static final RegistrySupplier<MenuType<RksMachineContainer>> RKS_MACHINE = register("rks_machine", RksMachineContainer::new, RksMachineBlockEntity.class);


    public static void init() {
        CONTAINERS.register();
        PlayerEvent.CLOSE_MENU.register(GenerationsContainers::onContainerClose);
    }

    private static void onContainerClose(Player player, AbstractContainerMenu container) {
        if (container instanceof MelodyFluteContainer melodyFluteContainer) melodyFluteContainer.save();
        if (container instanceof WalkmonContainer genericChestContainer) genericChestContainer.save();
    }

    public static <T extends AbstractContainerMenu, V extends BlockEntity> RegistrySupplier<MenuType<T>> register(String name, Function<CreationContext<V>, T> function, Class<V> clazz) {
        return CONTAINERS.register(name, () -> MenuRegistry.ofExtended((id, playerInventory, arg2) -> {
            var be = playerInventory.player.level().getBlockEntity(arg2.readBlockPos());
            if (clazz.isInstance(be)) return function.apply(new CreationContext<>(id, playerInventory, clazz.cast(be)));
            else return null;
        }));
    }

    public record CreationContext<V extends BlockEntity>(int id, Inventory playerInv, V blockEntity) {}
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/container/GenerationsContainers.kt
package generations.gg.generations.core.generationscore.common.world.container

import generations.gg.generations.core.generationscore.common.GenerationsCore
import generations.gg.generations.core.generationscore.common.util.PlatformRegistry
import net.minecraft.core.Holder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.flag.FeatureFlags
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.MenuType

object GenerationsContainers: PlatformRegistry<MenuType<*>>(Registries.MENU, BuiltInRegistries.MENU) {

    @JvmField
    val COOKING_POT = register("cooking_pot", ::CookingPotContainer)
//    val MACHINE_BLOCK = register("machine_block", ::MachineBlockContainer, MachineBlockEntity::class.java)
    val MELODY_FLUTE = register("melody_flute", ::MelodyFluteContainer)
    val TRASHCAN = register("trashcan", ::TrashCanContainer)
    val RKS_MACHINE = register("rks_machine", ::RksMachineContainer)
    val GENERIC = registerExtended("generic", GenericChestContainer.Companion::fromBuffer)


    fun <T: AbstractContainerMenu> registerExtended(name: String, constructor: (Int, Inventory, FriendlyByteBuf) -> T): Holder<MenuType<T>> = create(name, GenerationsCore.implementation.createExtendedMenu<T>(constructor))

    override fun init() {
        super.init()

//        PlayerEvent.CLOSE_MENU.register(PlayerEvent.CloseMenu { player: Player, container: AbstractContainerMenu ->
//            onContainerClose(
//                player, container
//            )
//        })
    }

    private fun onContainerClose(player: Player, container: AbstractContainerMenu) {
//        if(container instanceof GenericChestContainer genericChestContainer) {
//            genericChestContainer.getContainer().update();
//        }

//        if (container instanceof WalkmonContainer genericChestContainer) genericChestContainer.save(player);
//        if (container instanceof CalyrexSteedContainer genericChestContainer) genericChestContainer.save(player);

//        if (container is RksMachineContainer) container.close()
    }

    fun <T: AbstractContainerMenu> register(name: String, constructor: (Int, Inventory) -> T): Holder<MenuType<T>> = create<MenuType<T>>(name) { MenuType(constructor::invoke, FeatureFlags.VANILLA_SET) }

//    fun <T : AbstractContainerMenu, V : BlockEntity> register(
//        name: String,
//        function: Function<CreationContext<V>, T>,
//        clazz: Class<V>
//    ): MenuType<T> {
//        return create(name.generationsResource(), GenerationsCore.implementation.createExtendedMenu { id: Int, playerInventory: Inventory, arg2: FriendlyByteBuf ->
//            val be = playerInventory.player.level().getBlockEntity(arg2.readBlockPos())
//            if (clazz.isInstance(be)) return@createExtendedMenu function.apply(
//                CreationContext<V>(
//                    id,
//                    playerInventory,
//                    clazz.cast(be)
//                )
//            )
//            else return@createExtendedMenu null
//        })
//    }


//    @JvmRecord
//    data class CreationContext<V : BlockEntity?>(val id: Int, val playerInv: Inventory, val blockEntity: V)
}
